### ISSUE TITLE
Decryption fails on particular track.

### ISSUE THREAD

**Issue by MarcG2** [redsudo/RedSea#44]
Program crashes when it tries to decrypt first track on:
https://listen.tidal.com/album/85187772

Quality is set to best available. Perhaps it's due to the file size?

**Comment by redsudo** [comment]
What region is your account in? I am having a hard time reproducing since my test credentials can't access that album due to region locking. Once I get account credentials for that region, I can test and fix.

**Comment by MarcG2** [comment]
I'm in the US.

**Comment by redsudo** [comment]
I tried using a US based premium account on the latest redsea release (2.2) and it worked:

> <<< Downloading single track... >>>
> === Downloading track ID 85187773 ===
>         Track: 01
>         Title: The Greatest Show on Earth (Remastered)
>         Artist: Nightwish
>         Album: Decades
>         ----
>         Grabbing album info...
>         Grabbing stream URL...
>         Download progress: 100%
>         Looks like file is encrypted. Decrypting...
>         Downloading album art...
>         Download progress: 100%
>         Tagging media file...
> === 1/1 complete (100% done) ===
> 
> Download queue: 1/1 items complete (100% done) <
> 
> All downloads completed. <

Can you send me the log output?

**Comment by MarcG2** [comment]
You are downloading at the highest quality right?  Right after it outputs "Looks like file is encrypted. Decrypting....", it crashes.  It doesn't output any error. 
```

 /$$$$$$$                  /$$  /$$$$$$
| $$__  $$                | $$ /$$__  $$
| $$  \ $$  /$$$$$$   /$$$$$$$| $$  \__/  /$$$$$$   /$$$$$$
| $$$$$$$/ /$$__  $$ /$$__  $$|  $$$$$$  /$$__  $$ |____  $$
| $$__  $$| $$$$$$$$| $$  | $$ \____  $$| $$$$$$$$  /$$$$$$$
| $$  \ $$| $$_____/| $$  | $$ /$$  \ $$| $$_____/ /$$__  $$
| $$  | $$|  $$$$$$$|  $$$$$$$|  $$$$$$/|  $$$$$$$|  $$$$$$$
|__/  |__/ \_______/ \_______/ \______/  \_______/ \_______/

                    (c) 2016 Joe Thatcher
               https://github.com/Azalius/RedSea


<<< Downloading album: 22 track(s) in total >>>
=== Downloading track ID 85187773 ===
        Track: 01
        Title: The Greatest Show on Earth (Remastered)
        Artist: Nightwish
        Album: Decades
        ----
        Grabbing stream URL...
        Download progress: 100%
        Looks like file is encrypted. Decrypting...
Exiting...
```

**Comment by redsudo** [comment]
Yes. Here is the command I used and the output:

> ~/D/G/R/RedSea> pipenv run python redsea.py -p best_available https://tidal.com/browse/track/85187773
> 
> <<< Downloading single track... >>>
> === Downloading track ID 85187773 ===
>         Track: 01
>         Title: The Greatest Show on Earth (Remastered)
>         Artist: Nightwish
>         Album: Decades
>         ----
>         Grabbing album info...
>         Grabbing stream URL...
>         Download progress: 100%
>         Looks like file is encrypted. Decrypting...
>         Downloading album art...
>         Download progress: 100%
>         Tagging media file...
> === 1/1 complete (100% done) ===
> 
> Download queue: 1/1 items complete (100% done) <
> 
> All downloads completed. <


Then I used mqaid to verify that it did indeed pull the MQA (highest quality) version:

> 
> ~/D/G/R/mqaid> pipenv run python is_mqa.py .
>  Found 1 flac file(s). Decoding now... 
>  MQA syncword present. [Nightwish - Decades/01 - The Greatest Show on Earth Remastered.flac]

Just a heads up, MQA format contains some artifacts (such as the syncword) that are generally unfavorable for certain applications vs standard 16bit flac but for personal use you should be fine.

**Comment by MarcG2** [comment]
I didn't realize the flac files I was getting also contained MQA data.  Is it just metadata or is it mixed in with the encode? 

I'm not on that IRC.  Are you using different versions of the dependencies? This is what I have.
mutagen       1.42.0
opencv-python 3.4.4.19
pycryptodome  3.8.2
requests      2.9.1

Is there a way to get the application give information on whatever exception is being thrown?

**Comment by redsudo** [comment]
Yeah, even though MQA is technically higher quality since it is 24bit, it does contain artifacts embedded in the audio which can only be properly removed with MQA compatible hardware (the software MQA decoder will not completely reverse the MQA encoding).

The embedded data is not supposed to be audible so like many things in audio, it's a matter of debate/personal opinion whether its better to have MQA or standard 16bit flac. That's why I have the software default to 16bit flac but retain the option to download MQA. Perhaps I should add a warning/note regarding the MQA stuff.

I am using these dependencies with standard Python 3.7.4:
```astroid==2.2.5
isort==4.3.21
lazy-object-proxy==1.4.2
mccabe==0.6.1
mutagen==1.31
pycryptodome==3.9.0
requests==2.9.1
six==1.12.0
typed-ast==1.4.0
wrapt==1.11.2```

Regarding increased verbosity, there isn't currently an option for that but I will add it to the to-do list. Still, the place where the program is halting for you is really strange. "Exiting.." is not one of the messages in this program so it must be from one of the libraries, likely the crypto library.

I will work on getting a more verbose version of the program together.

**Comment by MarcG2** [comment]
I tried downloading the album at FLAC quality instead, and it completed successfully.  So the issue seems to be with the MQA files.
